File: core/src/main/java/io/github/fourlastor/game/demo/round/ability/PoisonAbility.java
package io.github.fourlastor.game.demo.round.ability;

import dagger.assisted.Assisted;
import dagger.assisted.AssistedFactory;
import dagger.assisted.AssistedInject;
import io.github.fourlastor.game.demo.round.Ability;
import io.github.fourlastor.game.demo.round.StateRouter;
import io.github.fourlastor.game.demo.round.UnitInRound;
import io.github.fourlastor.game.demo.round.step.StepState;
import io.github.fourlastor.game.demo.round.step.Steps;
import io.github.fourlastor.game.demo.state.Filter;
import io.github.fourlastor.game.demo.state.GameState;
import io.github.fourlastor.game.demo.state.map.Tile;
import io.github.fourlastor.game.demo.state.unit.Unit;
import java.util.function.BiPredicate;
import java.util.function.Predicate;
import space.earlygrey.simplegraphs.algorithms.SearchStep;

public class PoisonAbility extends Ability {

    private final Unit unit;
    private final Steps steps;

    @AssistedInject
    public PoisonAbility(
            @Assisted UnitInRound unitInRound, StateRouter router, StepState.Factory stateFactory, Steps steps) {
        super(unitInRound, router, stateFactory);
        this.unit = unitInRound.unit;
        this.steps = steps;
    }

    @Override
    protected Builder<?> createSteps(GameState state) {
        Predicate<SearchStep<Tile>> movementLogic =
                Filter.all(Filter.maxDistance(unit.type.speed + 1), Filter.canTravel(unit));
        BiPredicate<GameState, Tile> searchLogic = Filter.all(
                Filter.canReach(state.tileAt(unit.hex), movementLogic),
                (unused, tile) -> state.unitAt(tile.hex) != null);
        return start(steps.searchTile(searchLogic)).then(hex -> steps.poison(unit, state.unitAt(hex)));
    }

    @AssistedFactory
    public interface Factory {
        PoisonAbility create(UnitInRound unitInRound);
    }
}

File: core/src/main/java/io/github/fourlastor/game/demo/round/ability/PoisonAbility.kt
package io.github.fourlastor.game.demo.round.ability

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import io.github.fourlastor.game.demo.round.Ability
import io.github.fourlastor.game.demo.round.StateRouter
import io.github.fourlastor.game.demo.round.UnitInRound
import io.github.fourlastor.game.demo.round.step.StepState
import io.github.fourlastor.game.demo.round.step.Steps
import io.github.fourlastor.game.demo.state.Filter.all
import io.github.fourlastor.game.demo.state.Filter.canReach
import io.github.fourlastor.game.demo.state.Filter.canTravel
import io.github.fourlastor.game.demo.state.Filter.maxDistance
import io.github.fourlastor.game.demo.state.GameState
import io.github.fourlastor.game.demo.state.unit.Mon
import io.github.fourlastor.game.demo.state.unit.effect.PoisonEffect
import java.util.function.BiPredicate

class PoisonAbility @AssistedInject constructor(
    @Assisted unitInRound: UnitInRound,
    router: StateRouter,
    stateFactory: StepState.Factory,
    private val steps: Steps,
    private val poisonEffect: PoisonEffect
) : Ability(unitInRound, router, stateFactory) {
    private val mon: Mon = unitInRound.mon

    override fun createSteps(state: GameState): Builder<*> {
        val movementLogic = all(maxDistance(mon.type.speed + 1), canTravel(mon))
        val searchLogic = all(
            canReach(state.tileAt(mon.hex), movementLogic),
            BiPredicate { _, tile -> state.unitAt(tile.hex) != null }
        )
        return start(steps.searchTile(searchLogic)).step { hex, continuation ->
            val target = requireNotNull(state.unitAt(hex))
            target.group.image.addAction(
                Actions.sequence(
                    Actions.color(Color.OLIVE, 0.5f),
                    Actions.color(Color.GREEN, 0.2f),
                    Actions.color(Color.WHITE, 0.3f),
                    Actions.run { target.addEffect(poisonEffect) },
                    Actions.run(continuation)
                )
            )
        }
    }

    @AssistedFactory
    interface Factory {
        fun create(unitInRound: UnitInRound): PoisonAbility
    }
}
